### PR TITLE
Add functionality for Pareto optimization

### PR DESCRIPTION
**Pull request by thomaswmorris**
Pareto optimization is an alternative to scalarization where we explore the Pareto front (https://en.wikipedia.org/wiki/Pareto_efficiency).

### CODE REVIEW

File: src/blop/agent.py [hunk 1]
```diff
@@ -542,32 +546,86 @@ def posterior(self, x):
...
+        # a point is on the Pareto front if it is Pareto dominant
+        # a point is Pareto dominant if it is there is no other point that is better at every objective
+        y = self.train_targets()[:, self.objectives.type == "fitness"]
+        in_pareto_front = ~(y.unsqueeze(1) > y.unsqueeze(0)).all(axis=-1).any(axis=0)
+        all_constraints_satisfied = self.evaluated_constraints.all(axis=-1)
+        return in_pareto_front & all_constraints_satisfied
+
+    @property
+    def pareto_front(self):
+        return self.table.loc[self.pareto_front_mask.numpy()]
```
Comment (mrakitin) [inline src/blop/agent.py:613]: I may need some explanation on the `pareto_front_mask` and `pareto_front`. Maybe a docstring can have that explanation?

File: src/blop/objectives.py [hunk 2]
```diff
@@ -197,13 +236,13 @@ def summary(self) -> pd.DataFrame:
...
-        return table
+        return table.T
 
     def __repr__(self):
         return self.summary.__repr__()
 
-    def __repr_html__(self):
-        return self.summary.__repr_html__()
+    def _repr_html_(self):
+        return self.summary._repr_html_()
```
[the remark this answers is not in the document]
Comment (mrakitin) [reply src/blop/objectives.py:245]: Found it: https://ipython.readthedocs.io/en/stable/config/integrating.html#custom-methods

File: src/blop/plotting.py [hunk 3]
```diff
@@ -3,10 +3,10 @@
 from matplotlib import pyplot as plt
 from matplotlib.patches import Patch
 
-from . import acquisition
+from .bayesian import acquisition
 
 DEFAULT_COLOR_LIST = ["dodgerblue", "tomato", "mediumseagreen", "goldenrod"]
-DEFAULT_COLORMAP = "magma"
+DEFAULT_COLORMAP = "turbo"
```
Comment (mrakitin) [inline src/blop/plotting.py:9]: What's better about this color map?

File: src/blop/plotting.py [hunk 4]
```diff
@@ -89,77 +89,131 @@ def _plot_objs_many_dofs(agent, axes=(0, 1), shading="nearest", cmap=DEFAULT_COL
...
+                    test_x,
+                    test_y,
+                    c=test_constraint,
+                    s=size,
+                    cmap=cmap,
+                    norm=mpl.colors.LogNorm(),
+                )
+
+        val_cbar = agent.obj_fig.colorbar(val_ax, ax=agent.obj_axes[obj_index, 0], location="bottom", aspect=32, shrink=0.8)
+        val_cbar.set_label(f"{obj.units if obj.units else ''}")
```
Comment (mrakitin) [inline src/blop/plotting.py:188]: It can be shorter:
```suggestion
        val_cbar.set_label(f"{obj.units or ''}")
```

**Review by mrakitin — APPROVED** [review]
Looks good overall. I posted a few comments and suggestions below.

I think what is missing is an example (Jupyter/Sphinx) to demonstrate the Pareto front optimization. Can it be added, please?

**Review by mrakitin — CHANGES_REQUESTED** [review]
Need an example. The "approving" review was preliminary.

File: src/blop/bayesian/transforms.py [hunk 5]
```diff
@@ -6,6 +6,8 @@
 from botorch.posteriors.posterior_list import PosteriorList
 from torch import Tensor
 
+# This file is deprecated for now, but we may need it later
+
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply src/blop/bayesian/transforms.py:10]: I will delete this actually, some of the pareto-conscious acquisition functions won't accept posterior transforms so we need to use the other workaround anyway.

File: src/blop/objectives.py [hunk 6]
```diff
@@ -233,6 +272,13 @@ def weights(self) -> np.array:
         """
         return np.array([obj.weight for obj in self.objectives])
 
+    @property
+    def is_fitness(self) -> np.array:
+        """
+        Returns an array of the objective weights.
+        """
+        return np.array([obj.target in ["min", "max"] for obj in self.objectives])
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply src/blop/objectives.py:280]: This is to distinguish between fitness-based objectives and constraint-based objectives. It is removed in a later PR and is controlled by `objective.kind`

File: src/blop/agent.py [hunk 7]
```diff
@@ -542,32 +546,86 @@ def posterior(self, x):
...
 
     @property
-    def targeting_transform(self):
-        return TargetingPosteriorTransform(
-            weights=torch.tensor(self.active_objs.weights, dtype=torch.double), targets=self.active_objs.targets
+    def fitness_model(self):
+        return (
+            ModelListGP(*[obj.model for obj in self.active_objs if obj.type == "fitness"])
+            if len(self.active_objs) > 1
+            else self.active_objs[0].model
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply src/blop/agent.py:553]: The `ModelList` object just stitches together the outputs, so it acts just like a model trained on multiple outputs (but this way the different objectives can have differently trained hyperparameters, which is important).

File: src/blop/objectives.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply src/blop/objectives.py:245]: This is just to make the pandas table look nicer in Jupyter

File: src/blop/plotting.py (hunk 3 above)
Comment (thomaswmorris) [reply src/blop/plotting.py:9]: The values near 1 are hard to see on a white background. Turbo goes from red to blue and isn't white in the middle.

Comment (mrakitin) [reply src/blop/plotting.py:9]: Can we add a comment/note about it next to the color map assignment?

Comment (mrakitin) [inline src/blop/plotting.py:9]:
```suggestion
# Note: the values near 1 are hard to see on a white background. Turbo goes from red to blue and isn't white in the middle.
DEFAULT_COLORMAP = "turbo"
```

File: src/blop/agent.py [hunk 8]
```diff
@@ -542,32 +546,93 @@ def posterior(self, x):
...
-        return np.argmax(np.where(np.isnan(f), -np.inf, f))
+    def best_f(self):
+        return self.scalarized_fitnesses[self.argmax_best_f]
+
+    @property
+    def pareto_front_mask(self):
+        """
+        A mask for all data points that is true when the point is on the Pareto front.
+        A point is on the Pareto front if it is Pareto dominant
+        A point is Pareto dominant if it is there is no other point that is better at every objective
```
Comment (mrakitin) [inline src/blop/agent.py:607]: To match the notebook example.
```suggestion
        A point is Pareto dominant if there is no other point that is better at every objective
```

**Review by mrakitin — APPROVED** [review]
Looks good, suggested a couple of minor corrections below, it's good to go after that.

Just to confirm - the learning process will still be the same, i.e. no special input is needed to declare we want to use the Pareto-efficient optimization. The plotting will help to understand the Pareto front and then one can pick some points from the front. Is my understanding correct?

**Comment by thomaswmorris** [comment]
> Looks good, suggested a couple of minor corrections below, it's good to go after that.
> 
> Just to confirm - the learning process will still be the same, i.e. no special input is needed to declare we want to use the Pareto-efficient optimization. The plotting will help to understand the Pareto front and then one can pick some points from the front. Is my understanding correct?

Yup, nothing is changed.

**Pull request merged**